internal/cmd/temporal-features-test-proxy/main.go

```diff
@@ -0,0 +1,752 @@
+		ctx := r.Context()
+		ctx = context.WithValue(ctx, queryKey, q)
+		if err := action(ctx); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.WriteHeader(http.StatusNoContent)
+	}
+}
+
+func CheckMethod(w http.ResponseWriter, r *http.Request, allowed ...string) bool {
+	for _, item := range allowed {
+		if r.Method == item {
+			return true
+		}
+	}
+	h := w.Header()
+	h.Set("Allow", strings.Join(allowed, ", "))
+	http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
+	return false
+}
+
+type ControlServer struct {
+	listen string
+	mu     sync.Mutex
+	cv     *sync.Cond
+	l      net.Listener
+	quitCh chan struct{}
+	server http.Server
+	mux    http.ServeMux
+}
+
+func (s *ControlServer) Init(listen string) {
+	s.listen = listen
+	s.cv = sync.NewCond(&s.mu)
+	s.l = nil
+	s.quitCh = nil
+	s.server.Handler = &s.mux
+	s.server.ReadTimeout = 30 * time.Second
+	s.server.WriteTimeout = 30 * time.Second
+	s.server.IdleTimeout = 60 * time.Second
+	s.mux.HandleFunc("/", HandleHelp)
+	s.mux.HandleFunc("/quit", HandleAction(ActionQuit))
+	s.mux.HandleFunc("/restart", HandleAction(ActionRestart))
+	s.mux.HandleFunc("/reject", HandleAction(ActionReject))
+	s.mux.HandleFunc("/accept", HandleAction(ActionAccept))
+	s.mux.HandleFunc("/freeze", HandleAction(ActionFreeze))
+	s.mux.HandleFunc("/thaw", HandleAction(ActionThaw))
+}
+
+func (s *ControlServer) Run(ctx context.Context) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.quitCh != nil {
+		panic("BUG! ControlServer is already running")
+	}
+
+	l, err := gListenConfig.Listen(ctx, "tcp", s.listen)
+	if err != nil {
+		return fmt.Errorf("failed to listen on %q: %w", s.listen, err)
+	}
+
+	s.l = l
+	s.quitCh = make(chan struct{})
+
+	s.server.BaseContext = func(l net.Listener) context.Context {
+		return gRootContext
+	}
+
+	go s.serveThread()
+	go s.closeThread(ctx, s.quitCh, &s.server)
+	return nil
+}
+
+func (s *ControlServer) Shutdown(ctx context.Context) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.quitCh == nil {
+		return nil
+	}
+
+	close(s.quitCh)
+	err := s.server.Shutdown(ctx)
+	for s.quitCh != nil {
+		s.cv.Wait()
+	}
+	return err
+}
+
+func (s *ControlServer) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.quitCh == nil {
+		return nil
+	}
+
+	close(s.quitCh)
+	err := s.server.Close()
+	for s.quitCh != nil {
+		s.cv.Wait()
+	}
+	return err
+}
+
+func (s *ControlServer) ForceClose() {
+	err := s.Close()
+	if IsErrClosed(err) {
+		err = nil
+	}
+	if err != nil {
+		Warn("failed to stop HTTP control server: %v", err)
+	}
+}
+
+func (s *ControlServer) closeThread(ctx context.Context, quitCh <-chan struct{}, closer io.Closer) {
+	select {
+	case <-ctx.Done():
+		err := closer.Close()
+		if IsErrClosed(err) {
+			err = nil
+		}
+		if err != nil {
+			Error("failed to stop HTTP control server: %v", err)
+		}
+	case <-quitCh:
+	}
+}
+
+func (s *ControlServer) serveThread() {
+	defer s.finish()
+
+	err := s.server.Serve(s.l)
```

Comment:
Big fan of using HTTP here. I wish when the "summary server" thing was done HTTP was used there instead of homemade Temporal binary protocol, heh.